include/rnp/rnp2.h

```diff
@@ -0,0 +1,241 @@
+/**
+* Export an unencrypted private key
+*/
+rnp_result rnp_private_export(rnp_private_key key,
+                              bool            armor,
+                              uint8_t **      output,
+                              size_t *        output_len);
+
+/**
+* Export an encrypted private key
+*/
+rnp_result rnp_private_export_encrypted(rnp_private_key key,
+                                        const char *    passphrase,
+                                        pgp_symm_alg_t  cipher,
+                                        pgp_hash_alg_t  s2k_hash,
+                                        uint32_t        s2k_hash_iterations,
+                                        uint8_t **      output,
+                                        size_t *        output_len);
+
+/* Signature/verification operations */
+
+enum rnp_signature_type { SIG_NORMAL, SIG_ARMOR, SIG_DETACHED };
+
+/**
+* Sign a message
+* @param key the private key, must be decrypted
+* @param hash the hash
+* @param type the signature type (binary, armored, or detached)
+* @param msg the message to sign
+* @param msg_len the length of msg in bytes
+* @param sig on success, the output signature buffer
+* @param sig_len on success, the length of sig in bytes
+*/
+rnp_result rnp_sign(rnp_private_key    key,
+                    pgp_hash_alg_t     hash,
+                    rnp_signature_type type,
+                    const uint8_t      msg[],
+                    size_t             msg_len,
+                    uint8_t **         sig,
+                    size_t *           sig_len);
+
+/**
+* Verify a signature with embedded message
+* @param key the public key
+* @param sig the signature
+* @param sig_len length of signature in bytes
+* @param msg on succes, output buffer to the message
+* @param msg_len on success, length of msg in bytes
+* @param hash_used if non-null and on success, set to the
+* hash algorithm used to create the signature.
+*/
+rnp_result rnp_verify(rnp_public_key  key,
+                      const uint8_t   sig[],
+                      size_t          sig_len,
+                      uint8_t **      msg,
+                      size_t *        msg_len,
+                      pgp_hash_alg_t *hash_used);
+
+/**
+* Verify a detached signature
+* @param key the key to verify with
+* @param msg the message bits being verified
+* @param msg_len length of msg in bytes
+* @parma sig the signature
+* @param sig_len length of signature in bytes
+* @param hash_used if non-null, and the signature is verified,
+* the hash algorithm used to create the signature will be returned.
+* @return RNP_SUCCESS if valid otherwise error
+*/
+rnp_result rnp_verify_detached(rnp_public_key  key,
+                               const uint8_t   msg[],
+                               size_t          msg_len,
+                               const uint8_t   sig[],
+                               size_t          sig_len,
+                               pgp_hash_alg_t *hash_used);
+
```

Comment:
This API may not be acceptable for detached verification. We should be able to validate a detached sig of a 1TB file, for example, without reading the entire thing into a memory buffer.

Reply:
I guess we could require clients to mmap, but I think it may be an awkward requirement to have some python/ruby script doing some platform-specific file memory mapping operation. Thoughts?

Reply:
Why not to specify file pathes/some kind of stream handles here? Will this work over FFI?

Reply:
Yeah good point, especially as it seems Ruby doesn't even have `mmap` built in. I think this function as it is would be useful for a lot of cases, but we'll need something else to deal with very large files.

We could pass a `FILE*` or file descriptor to read from. But file descriptor is Unix specific, and I'm not sure if creating a stdio `FILE*` for passing to a C function exists in Ruby or Python (in a brief search I couldn't find anything on this for either lang).

So I guess file path it is?

Reply:
> We could pass a FILE* or file descriptor to read from. But file descriptor is Unix specific, and I'm not sure if creating a stdio FILE* for passing to a C function exists in Ruby or Python (in a brief search I couldn't find anything on this for either lang).

It's not hard to use `fopen` [via ruby-ffi](https://github.com/ffi/ffi/blob/c52ceb173288073a321a94b401243ff40110d910/spec/ffi/memorypointer_spec.rb#L74) and I've done it in the past, but I would have concerns about inadvertently passing a `FILE*` created in one libc over to a library that may be using a different libc. Like what if my ruby code makes a reasonable assumption that `libc` is being used, so it uses `fopen` from `libc`...but it turns out `librnp` is using `musl`. I wouldn't expect it's standardized enough to do anything but crash, though I admit I haven't looked.

> So I guess file path it is?

Seems like it may be the simplest way.

Reply:
I think file path is the preferred approach, but we also need to make sure the FFI libraries handle files streaming properly, otherwise the improvements @ni4 made in memory usage could be annulled, say if the ruby-rnp reads in the whole file and passes it to librnp. Maybe some documentation could help.